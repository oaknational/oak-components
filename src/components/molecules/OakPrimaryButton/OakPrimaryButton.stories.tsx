import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakPrimaryButton } from "./OakPrimaryButton";

import { oakIconNames } from "@/components/atoms/OakIcon";
import { OakFlex } from "@/components/atoms";

const controlIconNames = [null, [...oakIconNames].sort()].flat();

const meta: Meta<typeof OakPrimaryButton> = {
  component: OakPrimaryButton,
  tags: ["autodocs"],
  title: "components/molecules/OakPrimaryButton",
  argTypes: {
    iconName: {
      options: controlIconNames,
      control: { type: "select" },
    },
    isTrailingIcon: {
      control: { type: "boolean" },
    },
    isLoading: {
      control: { type: "boolean" },
    },
  },
  parameters: {
    controls: {
      include: ["iconName", "isTrailingIcon", "isLoading", "type"],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakPrimaryButton>;

export const Default: Story = {
  render: (args) => (
    <OakFlex $gap="space-between-m">
      <OakPrimaryButton {...args}>Primary Button</OakPrimaryButton>
      <OakPrimaryButton {...args} disabled>
        Disabled Button
      </OakPrimaryButton>
      <OakPrimaryButton {...args} isLoading>
        Loading Button
      </OakPrimaryButton>
    </OakFlex>
  ),
  args: {
    iconName: "arrow-right",
  },
};

export const LinkStyledAsButton: Story = {
  render: (args) => (
    <OakFlex $gap="space-between-m">
      <OakPrimaryButton {...args}>Primary Link</OakPrimaryButton>
    </OakFlex>
  ),
  args: {
    element: "a",
    href: "/",
    iconName: "arrow-right",
  },
};
