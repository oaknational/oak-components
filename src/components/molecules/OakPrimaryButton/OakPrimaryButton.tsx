import React, { ElementType } from "react";

import {
  InternalShadowRectButton,
  InternalShadowRectButtonProps,
} from "@/components/molecules/InternalShadowRectButton";
import { PolymorphicPropsWithoutRef } from "@/components/polymorphic";

export type OakPrimaryButtonProps = Omit<
  InternalShadowRectButtonProps,
  | "defaultBorderColor"
  | "defaultBackground"
  | "defaultTextColor"
  | "hoverBackground"
  | "hoverBorderColor"
  | "hoverTextColor"
  | "disabledBackground"
  | "disabledBorderColor"
  | "disabledTextColor"
>;

/**
 *
 * A specific implementation of InternalRectButton
 *
 * The following callbacks are available for tracking focus events:
 *
 * ### onClick
 * `onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;`
 *
 * ### onHovered
 *  `onHovered?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>, duration: number) => void;`<br>
 *  called after a mouseEnter and mouseLeave event has happened
 */
export const OakPrimaryButton = <C extends ElementType = "button">({
  element,
  ...rest
}: OakPrimaryButtonProps & PolymorphicPropsWithoutRef<C>) => {
  return (
    <InternalShadowRectButton
      element={element ?? "button"}
      defaultBorderColor="bg-btn-primary"
      defaultBackground="bg-btn-primary"
      defaultTextColor="text-inverted"
      hoverBackground="bg-btn-primary-hover"
      hoverBorderColor="bg-btn-primary-hover"
      hoverTextColor="text-inverted"
      disabledBackground="bg-btn-primary-disabled"
      disabledBorderColor="text-disabled"
      disabledTextColor="text-inverted"
      {...rest}
    />
  );
};
