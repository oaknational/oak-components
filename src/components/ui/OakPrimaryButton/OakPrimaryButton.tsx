import React from "react";

import {
  InternalRectButton,
  InternalRectButtonProps,
} from "@/components/ui/InternalRectButton";

export type OakPrimaryButtonProps = Pick<
  InternalRectButtonProps,
  | "children"
  | "className"
  | "data-testid"
  | "onClick"
  | "onHovered"
  | "disabled"
  | "isLoading"
  | "iconName"
  | "isTrailingIcon"
>;

export const OakPrimaryButton = (props: OakPrimaryButtonProps) => {
  return (
    <InternalRectButton
      $background="white"
      $color="black"
      $ba="border-solid-m"
      defaultBorderColor="bg-btn-primary"
      defaultBackground="bg-btn-primary"
      defaultTextColor="text-inverted"
      hoverBackground="bg-btn-primary-hover"
      hoverBorderColor="bg-btn-primary-hover"
      hoverTextColor="text-inverted"
      disabledBackground="bg-btn-primary-disabled"
      disabledBorderColor="text-disabled"
      disabledTextColor="text-inverted"
      {...props}
    />
  );
};
