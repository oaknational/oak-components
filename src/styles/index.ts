export * from "@/styles/theme/color";
