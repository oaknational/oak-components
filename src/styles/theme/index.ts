export { oakColorTokens, oakUiRoleTokens } from "./color";

export type {
  OakColorToken,
  OakCombinedColorToken,
  OakUiRoleToken,
  OakColorFilterToken,
} from "./color";

export {
  oakAllSpacingTokens,
  oakInnerPaddingTokens,
  oakSpaceBetweenTokens,
} from "./spacing";

export type {
  OakAllSpacingToken,
  OakCombinedSpacingToken,
  OakInnerPaddingToken,
  OakSpaceBetweenToken,
} from "./spacing";

export { oakDropShadowTokens } from "./dropShadow";
export type { OakDropShadowToken } from "./dropShadow";

export { oakOpacityTokens } from "./opacity";
export type { OakOpacityToken } from "./opacity";

export { oakTransitionTokens } from "./transitions";
export type { OakTransitionToken } from "./transitions";

export { oakBorderRadiusTokens, oakBorderWidthTokens } from "./borders";
export type { OakBorderRadiusToken, OakBorderWidthToken } from "./borders";

export type { OakTheme } from "./theme";

export { oakFontTokens, oakFontSizeTokens } from "./typography";
export type { OakFontToken, OakFontSizeToken } from "./typography";

export { oakZIndexTokens } from "./zIndex";

export { oakDefaultTheme } from "./default.theme";
