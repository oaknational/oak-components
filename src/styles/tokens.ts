export const oakAllSpacingPx = {
  "all-spacing-1": 4,
  "all-spacing-2": 8,
  "all-spacing-3": 12,
  "all-spacing-4": 16,
  "all-spacing-5": 20,
  "all-spacing-6": 24,
  "all-spacing-7": 32,
  "all-spacing-8": 40,
  "all-spacing-9": 48,
  "all-spacing-10": 56,
  "all-spacing-11": 64,
  "all-spacing-12": 72,
  "all-spacing-13": 80,
  "all-spacing-14": 92,
  "all-spacing-15": 100,
  "all-spacing-16": 120,
};

export type OakAllSpacing = keyof typeof oakAllSpacingPx;

export const oakInnerPaddingAllSpacing = {
  "inner-padding-xs": "all-spacing-2",
  "inner-padding-s": "all-spacing-3",
  "inner-padding-m": "all-spacing-4",
  "inner-padding-l": "all-spacing-5",
  "inner-padding-xl": "all-spacing-6",
};

export type OakInnerPadding = keyof typeof oakInnerPaddingAllSpacing;

export const oakSpaceBetweenAllSpacing = {
  "space-between-ssx": "all-spacing-2",
  "space-between-xs": "all-spacing-3",
  "space-between-s": "all-spacing-4",
  "space-between-m": "all-spacing-6",
  "space-between-m2": "all-spacing-7",
  "space-between-l": "all-spacing-9",
  "space-between-xl": "all-spacing-10",
  "space-between-xxl": "all-spacing-12",
  "space-between-xxxl": "all-spacing-13",
};

export type OakSpaceBetween = keyof typeof oakSpaceBetweenAllSpacing;

export const oakAllColorsHex = {
  white: "#ffffff",
  grey20: "#f2f2f2",
  grey30: "#e4e4e4",
  grey40: "#cacaca",
  grey50: "#808080",
  grey60: "#575757",
  grey70: "#2d2d2d",
  black: "#222222",
  oakGreen: "#287c34",
  mint: "bef2bd",
  mint30: "ebfbeb",
  mint50: "dff9de",
  aqua: "#b0e2de",
  aqua30: "#e7f6f5",
  aqua50: "#cee7e5",
  lavender: "#a0b6f2",
  lavender30: "#e3e9fb",
  lavender50: "#c6d1ef",
  pink: "#deb7d5",
  pink30: "#f5e9f2",
  pink50: "#e5d1e0",
  lemon: "#ffe555",
  lemon30: "#fff7cc",
  lemon50: "#f6e8a0",
  amber: "#ff934e",
  amber30: "#ffdfca",
  amber50: "#ffc8a6",
  red: "#dd0035",
  navy: "#0d24c4",
  blue: "#374cf1",
  magenta: "#d02aa7",
  purple: "#845ad9",
  teal: "#037b7d",
};

export type OakColor = keyof typeof oakAllColorsHex;

export const oakColorTokens = [
  "text-primary",
  "text-subdued",
  "text-error",
  "text-disabled",
  "text-link-active",
  "text-link-hover",
  "text-link-visited",
  "text-inverted",
  "text-success",
  "text-warning",
  "bg-primary",
  "bg-neutral",
  "bg-neutral-50",
  "bg-neutral-30",
  "bg-btn-primary",
  "bg-btn-primary-hover",
  "bg-btn-primary-disabled",
  "bg-btn-secondary",
  "bg-btn-secondary-hover",
  "bg-btn-secondary-disabled",
  "bg-icon",
  "bg-icon-hover",
  "bg-decorative1â€“main",
  "bg-decorative1-subdued",
  "bg-decorative1-very subdued",
  "bg-decorative2-main",
  "bg-decorative2-subdued",
  "bg-decorative2-very-subdued",
  "bg-decorative3",
  "bg-decorative3-50",
  "bg-decorative3-30",
  "bg-decorative4",
  "bg-decorative4-50",
  "bg-decorative4-30",
  "bg-decorative5",
  "bg-decorative5-50",
  "bg-decorative5-30",
  "bg-decorative6",
  "bg-decorative6-50",
  "bg-decorative6-30",
  "border-primary",
  "border-inverted",
  "border-neutral",
  "border-decorative1",
  "border-decorative2",
  "border-decorative3",
  "border-decorative4",
  "border-decorative5",
  "border-decorative6",
] as const;

export type OakColorToken = (typeof oakColorTokens)[number];
