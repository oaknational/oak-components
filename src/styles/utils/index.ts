export * from "./colorStyle";
export * from "./positionStyle";
export * from "./sizeStyle";
export * from "./spacingStyle";
export * from "./flexStyle";
