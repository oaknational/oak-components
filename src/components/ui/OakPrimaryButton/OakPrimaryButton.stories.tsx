import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakPrimaryButton } from "./OakPrimaryButton";

import { oakIconNames } from "@/components/base/OakIcon";
import { OakFlex } from "@/components/base";

const controlIconNames = [null, [...oakIconNames].sort()].flat();

/**
 *
 * A specific implementation of InternalRectButton
 *
 * The following callbacks are available for tracking focus events:
 *
 * ### onClick
 * `onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;`
 *
 * ### onHovered
 *  `onHovered?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>, duration: number) => void;`<br>
 *  called after a mouseEnter and mouseLeave event has happened
 */

const meta: Meta<typeof OakPrimaryButton> = {
  component: OakPrimaryButton,
  tags: ["autodocs"],
  title: "components/ui/OakPrimaryButton",
  argTypes: {
    iconName: {
      options: controlIconNames,
      control: { type: "select" },
    },
    isTrailingIcon: {
      control: { type: "boolean" },
    },
    isLoading: {
      control: { type: "boolean" },
    },
  },
  parameters: {
    controls: {
      include: ["iconName", "isTrailingIcon", "isLoading", "type"],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakPrimaryButton>;

export const Default: Story = {
  render: (args) => (
    <OakFlex $gap="space-between-m">
      <OakPrimaryButton {...args}>Primary Button</OakPrimaryButton>
      <OakPrimaryButton {...args} disabled>
        Disabled Button
      </OakPrimaryButton>
      <OakPrimaryButton {...args} isLoading>
        Loading Button
      </OakPrimaryButton>
    </OakFlex>
  ),
  args: {
    iconName: "arrow-right",
  },
};
