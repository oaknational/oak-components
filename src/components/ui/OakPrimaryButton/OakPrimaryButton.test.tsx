import React from "react";
import "@testing-library/jest-dom";
import { create } from "react-test-renderer";
import { ThemeProvider } from "styled-components";

import { OakPrimaryButton, OakPrimaryButtonProps } from "./OakPrimaryButton";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { oakDefaultTheme } from "@/styles";

const defaultArgs: OakPrimaryButtonProps = {
  iconName: "arrow-right",
};

describe("OakPrimaryButton", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it("renders", () => {
    const { getByTestId } = renderWithTheme(
      <OakPrimaryButton {...defaultArgs} data-testid="test">
        Click
      </OakPrimaryButton>,
    );
    expect(getByTestId("test")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = create(
      <ThemeProvider theme={oakDefaultTheme}>
        <OakPrimaryButton {...defaultArgs}>Click Me</OakPrimaryButton>
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders the chidren", () => {
    const { getByText } = renderWithTheme(
      <OakPrimaryButton {...defaultArgs}>Click</OakPrimaryButton>,
    );
    expect(getByText("Click")).toBeInTheDocument();
  });

  it("calls onClick method", () => {
    const onClick = jest.fn();
    const { getByTestId } = renderWithTheme(
      <OakPrimaryButton {...defaultArgs} data-testid="test" onClick={onClick}>
        Click
      </OakPrimaryButton>,
    );
    getByTestId("test").click();
    expect(onClick).toHaveBeenCalled();
  });
});
