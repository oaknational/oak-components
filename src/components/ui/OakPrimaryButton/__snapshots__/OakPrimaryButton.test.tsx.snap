// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakPrimaryButton matches snapshot 1`] = `
<div
  className="sc-aXZVg sc-eDPEul cKKnNr hPNaOd button-wrapper"
>
  <div
    className="sc-aXZVg gEUKeC grey-shadow"
  />
  <div
    className="sc-aXZVg gEUKeC yellow-shadow"
  />
  <button
    className="sc-cwHptR sc-dLMFU evvJac kzUrLL internal-button"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="sc-aXZVg sc-gEvEer goWQoj OzInM"
    >
      <div
        className="sc-aXZVg sc-gsFSXq dTLJnG iRYSeg"
      >
        <img
          alt="arrow-right"
          className="sc-dcJsrY eRtzRJ"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="https://undefined/undefined/v1699895398/icons/fv0z57zerrioft52dd9n.svg"
          style={
            {
              "bottom": 0,
              "color": "transparent",
              "height": "100%",
              "left": 0,
              "objectFit": undefined,
              "objectPosition": undefined,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": "100%",
            }
          }
        />
      </div>
      <span
        className="sc-eqUAAy gXELRB"
      >
        Click Me
      </span>
    </div>
  </button>
</div>
`;
